### PR TITLE
Adds support for ID, Class & Tag selectors to Element.querySelector

### PR DESCRIPTION
**Pull request by nainar**
Addresses #111

### CODE REVIEW

**Comment by nainar** [comment]
@kristoferbaxter PTAL? Can't assign you as a reviewer.

File: src/worker-thread/dom/Element.ts [hunk 1]
```diff
@@ -386,5 +386,28 @@ export class Element extends Node {
   public getElementsByTagName(tagName: string): Element[] {
     return matchChildrenElements(this, tagName === '*' ? _ => true : element => element.tagName === tagName);
   }
+
+  /**
+   * @see https://developer.mozilla.org/en-US/docs/Web/API/Element/querySelector
+   * @return Element with matching selector.
```
Comment (kristoferbaxter) [inline src/worker-thread/dom/Element.ts]: Can you add a JS Doc comment for the argument `selector`?

File: src/worker-thread/dom/Element.ts [hunk 2]
```diff
@@ -386,5 +386,28 @@ export class Element extends Node {
...
+   */
+  public querySelector(selector: string): Element | null {
+    if (selector[0] == '#') {
```
Comment (kristoferbaxter) [inline src/worker-thread/dom/Element.ts]: Is there a reason for loose equality checking here?

File: src/worker-thread/dom/Element.ts [hunk 3]
```diff
@@ -386,5 +386,28 @@ export class Element extends Node {
...
+      return this.ownerDocument.getElementById(selector.substr(1));
+    } else if (selector[0] == '.') {
+      const matches = this.getElementsByClassName(selector.substr(1));
```
Comment (kristoferbaxter) [inline src/worker-thread/dom/Element.ts]: According to MDN `querySelector` uses the entire document's elements to find matches, then limits to the scope of the element requested from. Are we sure the behaviour is the same here if we intentionally limit to the `baseElement` the query selection was requested from?

[MDN Document](https://developer.mozilla.org/en-US/docs/Web/API/Element/querySelector)
> The entire hierarchy of elements is considered when matching, including those outside the set of elements including baseElement and its descendants

Just asking for my own knowledge, and perhaps to push for a comment once we have an explanation.

File: src/worker-thread/dom/Element.ts (hunk 1 above)
Comment (nainar) [reply src/worker-thread/dom/Element.ts]: Done.

File: src/worker-thread/dom/Element.ts [hunk 4]
```diff
@@ -386,5 +386,28 @@ export class Element extends Node {
...
+      if (matches) {
+        return matches[0];
+      }
+    } else {
+      selector = selector.toLowerCase();
+      const matches = this.getElementsByTagName(selector);
```
[the remark this answers is not in the document]
Comment (nainar) [reply src/worker-thread/dom/Element.ts]: Done.

File: src/worker-thread/dom/Element.ts (hunk 2 above)
Comment (nainar) [reply src/worker-thread/dom/Element.ts]: Nope. My mistake. Fixed.

File: src/worker-thread/dom/Element.ts (hunk 3 above)
Comment (nainar) [reply src/worker-thread/dom/Element.ts]: Fixed this to first check the document and then filter the results.

**Comment by kristoferbaxter** [comment]
@nainar – Changes from feedback look great! Want to merge this as is and add further selector support in a followup PR?

**Comment by nainar** [comment]
Yup, that's the intention :)

**Pull request merged**